mailstore-connector/src/com/axonivy/connector/mailstore/MailStoreService.java

```diff
@@ -0,0 +1,618 @@
+	 * called. It is not a problem, to call the close method on a closed object again.
+	 */
+	public static class MessageIterator implements Iterator<Message>, AutoCloseable {
+		private Store store;
+		private Folder srcFolder;
+		private Folder dstFolder;
+		private boolean delete;
+		private Message[] messages;
+		private int nextIndex;
+		private ClassLoader originalClassLoader;
+
+		private MessageIterator(String storeName, String srcFolderName, String dstFolderName, boolean delete, Predicate<Message> filter) {
+			try {
+				// Use own classloader so that internal classes of javax.mail API are found.
+				// If they cannot be found on the classpath, then mail content will not
+				// be recognized and always be reported as IMAPInputStream
+				originalClassLoader = Thread.currentThread().getContextClassLoader();
+				Thread.currentThread().setContextClassLoader(Session.class.getClassLoader());
+
+				this.delete = delete;
+				store = MailStoreService.openStore(storeName);
+				srcFolder = MailStoreService.openFolder(store, srcFolderName, Folder.READ_WRITE);
+				if(StringUtils.isNotBlank(dstFolderName)) {
+					dstFolder = MailStoreService.openFolder(store, dstFolderName, Folder.READ_WRITE);
+				}
+				messages = srcFolder.getMessages();
+
+				// pre-fetch headers
+				FetchProfile fetchProfile = new FetchProfile();
+				fetchProfile.add(FetchProfile.Item.ENVELOPE);
+				srcFolder.fetch(messages, fetchProfile);
+
+				if(filter != null) {
+					messages = Stream.of(messages).filter(filter).toArray(Message[]::new);
+				}
+
+				LOG.debug("Received {0}{1} messages.", messages.length, filter != null ? " matching" : "");
+
+				nextIndex = 0;
+			} catch(Exception e) {
+				close();
+				throw buildError("iterator").withCause(e).build();
+			}
+		}
+
+		/**
+		 * Close and sync all actions to the mail server.
+		 * 
+		 * Will be called automatically after the last element is fetched.
+		 * Must be called if the iterator does not run to the end.
+		 */
+		@Override
+		public void close() {
+			Exception exception = null;
+			if(dstFolder != null && dstFolder.isOpen()) {
+				try {
+					dstFolder.close();
+				} catch (Exception e) {
+					LOG.error("Could not close destination folder {0}", e, dstFolder);
+					if(exception == null) {
+						exception = e;
+					}
+				}
+			}
+			if(srcFolder != null && srcFolder.isOpen()) {
+				try {
+					srcFolder.close();
+				} catch (Exception e) {
+					LOG.error("Could not close source folder {0}", e, srcFolder);
+					if(exception == null) {
+						exception = e;
+					}
+				}
+			}
+			if(store != null) {
+				try {
+					store.close();
+				} catch (Exception e) {
+					LOG.error("Could not close store {0}", e, srcFolder);
+					if(exception == null) {
+						exception = e;
+					}
+				}
+			}
+			if(exception != null) {
+				buildError("close").withCause(exception);
+			}
+			Thread.currentThread().setContextClassLoader(originalClassLoader);

```

Comment:
sollte in einem finally block gestehen.

Reply:
*stehen

Reply:
try {
 // all other code from the close logic
} finally {
 Thread.currentThread().setContextClassLoader(originalClassLoader);
}

Reply:
Ok, mache ich!